lib/NeuraDialect/Mapping/HeuristicMapping/HeuristicMapping.cpp

```diff
@@ -194,5 +196,189 @@ bool HeuristicMapping::mapWithBacktrack(
+                   << entry.first.resource->getId()
+                   << " is already occupied by operation: " << *(entry.second)
+                   << "\n";
+      occupied_tiles.insert(dyn_cast<Tile>(entry.first.resource));
+    }
+  }
+
+  for (Tile *tile : architecture.getAllTiles()) {
+    // Skip tiles that are already allocated to other operations
+    if (occupied_tiles.count(tile)) {
+      llvm::outs() << "[caculateSpatialAward] Tile: " << tile->getId()
+                   << " is already occupied, skip it.\n";
+      continue;
+    }
+
+    if (!tile->canSupportOperation(getOperationKindFromMlirOp(op))) {
+      llvm::errs() << "[caculateSpatialAward] Tile: " << tile->getId()
+                   << " does not support operation: " << *op << "\n";
+      continue; // Skip tiles that cannot support the operation.
+    }
+
+    int earliest_start_time_step = target_level;
+
+    for (Operation *producer : producers) {
+      std::vector<MappingLoc> producer_locs =
+          mapping_state.getAllLocsOfOp(producer);
+      assert(!producer_locs.empty() && "No locations found for producer");
+
+      MappingLoc producer_loc = producer_locs.back();
+      earliest_start_time_step =
+          std::max(earliest_start_time_step, producer_loc.time_step + 1);
+    }
+
+    int latest_end_time_step = earliest_start_time_step + mapping_state.getII();
+    std::vector<MappingLoc> backward_users_locs;
+    for (Operation *user : backward_users) {
+      std::vector<MappingLoc> user_locs = mapping_state.getAllLocsOfOp(user);
+      assert(!user_locs.empty() && "No locations found for backward user");
+
+      MappingLoc backward_user_loc = user_locs.back();
+      latest_end_time_step =
+          std::min(latest_end_time_step,
+                   backward_user_loc.time_step + mapping_state.getII());
+      llvm::outs() << "[caculateSpatialAward] Latest end time step for " << *op
+                   << " on tile: " << tile->getId()
+                   << " is now: " << latest_end_time_step << "\n";
+      backward_users_locs.push_back(backward_user_loc);
+    }
+
+    int award = 2 * mapping_state.getII();
+    if (critical_ops.count(op)) {
+      award += tile->getDstTiles().size();
+      award += op->getOperands().size() -
+               getPhysicalHops(producers, tile, mapping_state);
+    }
+
+    if (!backward_users.empty()) {
+      llvm::outs() << "[caculateSpatialAward] earliest_start_time_step: "
+                   << earliest_start_time_step
+                   << ", latest_end_time_step: " << latest_end_time_step
+                   << " for tile: " << tile->getType() << "#" << tile->getId()
+                   << "\n";
+    }
+
+    for (int t = earliest_start_time_step; t < latest_end_time_step; t += 1) {
+      MappingLoc tile_loc_candidate = {tile, t};
+
+      // If the tile at time `t` is available, we can consider it for mapping.
+      if (mapping_state.isAvailableAcrossTime(tile_loc_candidate)) {
```

Comment:
Where we add constraints for:
 - Only one op can be supported by one tile across all time?
 - Such constraint should also be added to links.
 - Both calculateAward() and placeAndRoute() should apply such constraint.